### PR TITLE
Feature/scheduler

### PR DESCRIPTION
**Pull request by Yuto-web**
### 変更点
- Learning Rate Scheduler のmodelクラスを追加
- callbacksに学習率描画用のクラスを追加

### 備考
- 学習率描画ですが、schedulerで変化した学習率のみ描画します（Optimizerでの学習率変化は反映されない）
- 動作確認済み、学習確認済み

### CODE REVIEW

**Comment by aiorhiroki** [comment]
schedulerはtrain_paramsの中にありそう。

File: farmer/domain/model/trainer_model.py [hunk 1]
```diff
@@ -5,10 +5,11 @@
 from .config_model import Config
 from .train_params_model import TrainParams
 from .image_loader_model import ImageLoader
+from .scheduler_model import LRScheduler
 
 
 @dataclasses.dataclass
-class Trainer(Config, ImageLoader):
+class Trainer(Config, ImageLoader, LRScheduler):
```
Comment (aiorhiroki) [inline farmer/domain/model/trainer_model.py:12]: TrainParamsの方に入るのかな？

Comment (Yuto-web) [reply farmer/domain/model/trainer_model.py]: train_paramsにLRSchedulerを入れてoptuna探索できるようにしました。

File: farmer/domain/tasks/set_train_env_task.py [hunk 2]
```diff
@@ -96,11 +84,26 @@ def set_train_params(train_params: dict) -> dict:
...
+            # result_dir/trial#/learning/
+            self.config.learning_path = self.config.learning_path.replace(
+                "/learning", f"/trial{trial.number}/learning")
+            self.config.model_path = self.config.model_path.replace(
+                "/model", f"/trial{trial.number}/model")
+            self.config.image_path = self.config.image_path.replace(
+                "/image", f"/trial{trial.number}/image")
+
+            # set train params to params setted by optuna
+            train_params_dict = copy.deepcopy(self.config.optuna_params)
```
Comment (aiorhiroki) [inline farmer/domain/tasks/set_train_env_task.py:99]: ここは、deepcopyしなくて良い気がする

File: farmer/domain/tasks/set_train_env_task.py [hunk 3]
```diff
@@ -48,25 +49,12 @@ def _do_set_cpu_gpu_devices_task(self, gpu: str):
...
-            trial_result_path, self.config.image_dir)
-        self.config.tfboard_path = os.path.join(
-            trial_result_path, self.config.tfboard_dir)
-
+    def _do_set_train_params_task(self, trial):
         def set_train_params(train_params: dict) -> dict:
             for key, val in train_params.items():
-                if isinstance(val, dict):
+                if not val:
+                    continue
```
Comment (aiorhiroki) [inline farmer/domain/tasks/set_train_env_task.py:56]: これもいらない気がする

**Review by aiorhiroki — COMMENTED** [review]
だいたい良さそう

**Comment by aiorhiroki** [comment]
@Yuto-web

File: farmer/domain/tasks/set_train_env_task.py (hunk 2 above)
Comment (Yuto-web) [reply farmer/domain/tasks/set_train_env_task.py:99]: deepcopyじゃないと、この後のset_train_params(train_params_dict)の処理でself.config.optuna_paramsの中身が変わってしまうのでdeepcopyにしています。

File: farmer/domain/tasks/set_train_env_task.py (hunk 3 above)
Comment (Yuto-web) [reply farmer/domain/tasks/set_train_env_task.py:56]: おっしゃる通り要らないので削除します。

File: farmer/domain/model/trainer_model.py [hunk 4]
```diff
@@ -85,11 +85,13 @@ def __post_init__(self):
         # For optuna analysis hyperparameter
         def _check_need_optuna(train_params: dict):
             for val in train_params.values():
-                if isinstance(val, list):
+                if not val:
+                    continue
```
[the remark this answers is not in the document]
Comment (Yuto-web) [reply farmer/domain/model/trainer_model.py:89]: 要らないので削除します。

File: farmer/domain/tasks/set_train_env_task.py [hunk 5]
```diff
@@ -96,11 +84,26 @@ def set_train_params(train_params: dict) -> dict:
...
+        if self.config.optuna:
+            self.config.trial_number = trial.number
+            self.config.trial_params = trial.params
+            # result_dir/trial#/learning/
+            self.config.learning_path = self.config.learning_path.replace(
+                "/learning", f"/trial{trial.number}/learning")
+            self.config.model_path = self.config.model_path.replace(
+                "/model", f"/trial{trial.number}/model")
+            self.config.image_path = self.config.image_path.replace(
+                "/image", f"/trial{trial.number}/image")
```
Comment (Yuto-web) [inline farmer/domain/tasks/set_train_env_task.py:96]: ここtrial0の下にtrial1の結果を作ってしまうバグ修正がマージされていなかったので、マージします

File: farmer/domain/tasks/set_train_env_task.py (hunk 2 above)
Comment (aiorhiroki) [reply farmer/domain/tasks/set_train_env_task.py:100]: optuna_paramは変わっても良いやつかと思ってたけど、どう？

Comment (Yuto-web) [reply farmer/domain/tasks/set_train_env_task.py:100]: 変わってしまうとまずいですね。optuna_paramsに各パラメータ―のlow, highが設定されているので。

Comment (Yuto-web) [reply farmer/domain/tasks/set_train_env_task.py:100]: optuna_params自体必要ないので、削除しました。

Comment (Yuto-web) [reply farmer/domain/tasks/set_train_env_task.py:100]: optuna_paramsは必要で、trainer_model.pyのdeepcopyが不要なので、deepcopy削除しました

**Review by aiorhiroki — COMMENTED** [review]
LGTM！
